src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/Pbkdf2Implementation.Windows.cs

```diff
@@ -0,0 +1,270 @@
+
+            // stackalloc 0 to let compiler know this cannot escape.
+            Span<byte> clearSpan = stackalloc byte[0];
+            ReadOnlySpan<byte> symmetricKeyMaterial = stackalloc byte[0];
+            int symmetricKeyMaterialLength;
+
+            if (password.IsEmpty)
+            {
+                // CNG won't accept a null pointer for the password.
+                symmetricKeyMaterial = stackalloc byte[1];
+                symmetricKeyMaterialLength = 0;
+                clearSpan = default;
+            }
+            else if (password.Length <= hashBlockSizeBytes)
+            {
+                // Password is small enough to use as-is.
+                symmetricKeyMaterial = password;
+                symmetricKeyMaterialLength = password.Length;
+                clearSpan = default;
+            }
+            else
+            {
+                // RFC 2104: "The key for HMAC can be of any length (keys longer than B bytes are
+                //     first hashed using H).
+                //     We denote by B the byte-length of such
+                //     blocks (B=64 for all the above mentioned examples of hash functions)
+                //
+                // Windows' PBKDF2 will do this up to a point. To ensure we accept arbitrary inputs for
+                // PBKDF2, we do the hashing ourselves.
+                Span<byte> hashBuffer = stackalloc byte[512 / 8]; // 64 bytes is SHA512, the largest digest handled.
+                int hashBufferSize;
+
+                switch (hashAlgorithmName)
+                {
+                    case HashAlgorithmNames.SHA1:
+                        hashBufferSize = SHA1.HashData(password, hashBuffer);
+                        break;
+                    case HashAlgorithmNames.SHA256:
+                        hashBufferSize = SHA256.HashData(password, hashBuffer);
+                        break;
+                    case HashAlgorithmNames.SHA384:
+                        hashBufferSize = SHA384.HashData(password, hashBuffer);
+                        break;
+                    case HashAlgorithmNames.SHA512:
+                        hashBufferSize = SHA512.HashData(password, hashBuffer);
+                        break;
+                    default:
+                        throw new CryptographicException();
+                }
+
+                clearSpan = hashBuffer.Slice(0, hashBufferSize);
+                symmetricKeyMaterial = clearSpan;
+                symmetricKeyMaterialLength = hashBufferSize;
+            }
+
+            Debug.Assert(symmetricKeyMaterial.Length > 0);
+
+            NTSTATUS generateKeyStatus;
+
+            if (s_usePseudoHandles)
+            {
+                fixed (byte* pSymmetricKeyMaterial = symmetricKeyMaterial)
+                {
+                    generateKeyStatus = Interop.BCrypt.BCryptGenerateSymmetricKey(
+                        (nuint)BCryptAlgPseudoHandle.BCRYPT_PBKDF2_ALG_HANDLE,
+                        out keyHandle,
```

Comment:
Just jotting down something that I thought of here.

This is the only meaningful allocation that occurs in the Windows 10 implementation. We _could_ avoid the allocation by forgoing the use of a safe handle and just use an `IntPtr` and free it ourselves. This could... reasonably... be done since the handle will never leave this method.

However, I was unable to determine any real world use case for the one shots where this allocation had any meaningful impact.